mlflow/utils/autologging_utils.py

```diff
@@ -275,3 +282,202 @@ def batch_metrics_logger(run_id):
     yield batch_metrics_logger
     batch_metrics_logger._purge()
+
+
+def autologging_integration(name):
+    """
+    Wraps an autologging function in order to store its configuration arguments. This enables
+    patch functions to broadly obey certain configurations (e.g., disable=True) without
+    requiring specific logic to be present in each autologging integration. **All autologging
+    integrations should be decorated with this wrapper.**
+    """
+
+    AUTOLOGGING_INTEGRATIONS[name] = {}
+
+    def wrapper(_autolog):
+        def autolog(*args, **kwargs):
+            AUTOLOGGING_INTEGRATIONS[name] = kwargs
+            _autolog(**kwargs)
+
+        wrapped_autolog = functools.wraps(_autolog)(autolog)
+        return wrapped_autolog
+
+    return wrapper
+
+
+def safe_patch(autologging_integration, destination, function_name, function):
+    """
+    Patches the specified `function_name` on the specified `destination` class for autologging
+    purposes, replacing its implementation with an error-safe copy of the specified `function`.
+
+    :param autologging_integration: The name of the autologging integration associated with the
+                                    patch.
+    :param destination: The Python class on which the patch function is being defined.
+    :param function_name: The name of the function to patch on the specified `destination` class.
+    :param function: The patch function to apply. The first argument to this function should be
+                     reserved for an `original` method argument representing the underlying /
+                     original function. Subsequent arguments should be identical to those of the
+                     original function being patched.
+    """
+
+    def patched_train(*args, **kwargs):
+        preexisting_run = mlflow.active_run()
+        original_result = None
+        called_original = False
+        failed_during_original = False
+
+        def call_original(*og_args, **og_kwargs):
+            original = gorilla.get_original_attribute(destination, function_name)
+
+            if _is_testing():
+                _validate_args(args, kwargs, og_args, og_kwargs)
+
+            def wrapped_original(*args, **kwargs):
+                try:
+                    called_original = True
+                    nonlocal original_result
+                    original_result = original(*args, **kwargs)
+                    return original_result
+                except Exception:
+                    nonlocal failed_during_original
+                    failed_during_original = True
+                    raise
+
+            return wrapped_original(*og_args, **og_kwargs)
+
+        original = gorilla.get_original_attribute(destination, function_name)
+        call_original = functools.wraps(original)(call_original)
+        call_original.__signature__ = inspect.signature(original)
+
+        config = AUTOLOGGING_INTEGRATIONS[autologging_integration]
```

Comment:
What are the set of configurations on autolog? How many of them can we handle in a common manner, or is 'disable' the only one?

Reply:
Currently, the common configurations are: `log_models`, `log_input_examples`, and `log_model_signatures`, and we intend to introduce `disable`. Of these configurations, `disable` seems like the easiest one to handle in a common manner. The other ones are trickier; we could provide utilities on top of `log_model` that respect these confs & have implementers use them, though this might be brittle.